pkg/controller/volume/populator/desired_state_of_world_populator.go

```diff
@@ -0,0 +1,112 @@
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
+)
+
+// DesiredStateOfWorldPopulator periodically verifies that the pods in the
+// desired state of th world still exist, if not, it removes them.
+// TODO: it also loops through the list of active pods and ensures that
+// each one exists in the desired state of the world cache
+// if it has volumes.
+type DesiredStateOfWorldPopulator interface {
+	Run(stopCh <-chan struct{})
+}
+
+// NewDesiredStateOfWorldPopulator returns a new instance of DesiredStateOfWorldPopulator.
+// loopSleepDuration - the amount of time the populator loop sleeps between
+//     successive executions
+// podManager - the kubelet podManager that is the source of truth for the pods
+//     that exist on this host
+// desiredStateOfWorld - the cache to populate
+func NewDesiredStateOfWorldPopulator(
+	loopSleepDuration time.Duration,
+	podInformer framework.SharedInformer,
+	desiredStateOfWorld cache.DesiredStateOfWorld) DesiredStateOfWorldPopulator {
+	return &desiredStateOfWorldPopulator{
+		loopSleepDuration:   loopSleepDuration,
+		podInformer:         podInformer,
+		desiredStateOfWorld: desiredStateOfWorld,
+	}
+}
+
+type desiredStateOfWorldPopulator struct {
+	loopSleepDuration   time.Duration
+	podInformer         framework.SharedInformer
+	desiredStateOfWorld cache.DesiredStateOfWorld
+}
+
+func (dswp *desiredStateOfWorldPopulator) Run(stopCh <-chan struct{}) {
+	wait.Until(dswp.populatorLoopFunc(), dswp.loopSleepDuration, stopCh)
+}
+
+func (dswp *desiredStateOfWorldPopulator) populatorLoopFunc() func() {
+	return func() {
+		dswp.findAndRemoveDeletedPods()
+	}
+}
+
+// Iterate through all pods in desired state of world, and remove if they no
+// longer exist in the informer
+func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
+	for dswPodUID, dswPodToAdd := range dswp.desiredStateOfWorld.GetPodToAdd() {
+		dswPodKey, err := kcache.MetaNamespaceKeyFunc(dswPodToAdd.Pod)
+		if err != nil {
+			glog.Errorf("Skip this pod, error occured in getting the namespace key from pod: %v", err)
+			continue
+		}
+		// retrieve the pod object from pod informer with the namespace key
+		informerPodObj, exists, err := dswp.podInformer.GetStore().GetByKey(dswPodKey)
+		if err != nil || informerPodObj == nil {
+			glog.Errorf("Skip this pod, error occured during checing pod existence from informer %v", err)
+			continue
+		}
+		if exists {
+			informerPod, ok := informerPodObj.(*api.Pod)
+			if !ok {
+				glog.Errorf("Failed to cast obj %#v to pod object for pod %q", informerPod, dswPodUID)
+				continue
+			}
+			informerPodUID := volumehelper.GetUniquePodName(informerPod)
+			// Check whether the unique idenfier of the pod from dsw matches the one retrived from pod informer
+			if informerPodUID == dswPodUID {
+				glog.V(10).Infof(
+					"The pod UID %q from dsw does match the one retrived from pod informer %v", dswPodUID, informerPod)
```

Comment:
`glog.V(10).Infof("Verified pod %q/%q (UID %q) from dsw exists in pod informer.", pod.Namespace, pod.Name, pod.UID)`